engine/consensus/approvals/approvals_lru_cache.go

```diff
@@ -0,0 +1,98 @@
+package approvals
+
+import (
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// LruCache is a wrapper over `simplelru.LRUCache` that provides needed api for processing result approvals
+// Extends functionality of `simplelru.LRUCache` by introducing additional index for quicker access.
+type LruCache struct {
+	lru  simplelru.LRUCache
+	lock sync.RWMutex
+	// secondary index by result id, since multiple approvals could
+	// reference same result
+	byResultID map[flow.Identifier]flow.IdentifierList
```

Comment:
I think we could use a map instead of a slice to store the approval list, this avoids duplicates.
```suggestion
	byResultID map[flow.Identifier]map[flow.Identifier]struct{}
```
`ByResultID` currently returns the IdentifierList, but I think we could either:
1. let it return a list of approvals
2. or replace with `TakeByResultID`, which is even better IMO